tests: posix: semaphores: ensure test is not skipped

cfriedt opened the pull request:
Commit f7633a55aa098de8409724f3a6211af4a4687101 moved the tests for the `POSIX_SEMAPHORES` Option Group from the `tests/posix/common` testsuite to `tests/posix/semaphores`.

However, there was a copy-paste error. Previously, tests would have been run only once when dynamic threads were enabled, and then skipped when dynamic threads were disabled, since that follows the posix programming model better. However, dynamic threads were never actually enabled after moving to the new testsuite. So all tests were effectively skipped.

Add the necessary options to prj.conf in order to ensure that there are sufficient dynamic threads available to run the testsuite.

Fixes #88768

Commit:
tests: posix: semaphores: ensure test is not skipped

Commit f7633a55aa098de8409724f3a6211af4a4687101 moved the tests for the
POSIX_SEMAPHORES Option Group from the tests/posix/common testsuite to
its own dedicated testsuite.

However, there was a copy-paste error. Previously, tests would have been
run only once when dynamic threads were enabled, and then skipped when
dynamic threads were disabled, since that follows the posix programming
model better. However, dynamic threads were never actually enabled after
moving to the new testsuite. So all tests were effectively skipped.

Add the necessary options to prj.conf in order to ensure that there are
sufficient dynamic threads available to run the testsuite.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

ycsin commented on tests/posix/semaphores/prj.conf:
```diff
@@ -3,3 +3,7 @@ CONFIG_ZTEST=y
 CONFIG_POSIX_SEMAPHORES=y
+
+CONFIG_DYNAMIC_THREAD=y
+CONFIG_DYNAMIC_THREAD_POOL_SIZE=2
```
Same as #88767, we can make it bulletproof by `BUILD_ASSERT`ing  `CONFIG_DYNAMIC_THREAD_POOL_SIZE>=2`

cfriedt replied:
Fixed!

cfriedt commented:
This needs another change unfortunately - it should be visible after a rebase, but due to #88538, the default heap-add with `CONFIG_POSIX_API` has been reduced from 1 kiB down to 256 bytes. Since posix semaphores allocate from the heap, and no effort was made to convert them to optionally use pooled allocators, `tests/posix/semaphores` fails after #88538 was merged.

I'll add another commit ahead of this one to create a heap-add for posix semaphores.

https://github.com/zephyrproject-rtos/zephyr/actions/runs/14535678361/job/40783625837?pr=88762

Commit:
posix: semaphores: use a default minimal heap-add for semaphores

The implementation of POSIX_SEMAPHORES historically used heap allocation
and has not yet been transitioned to a pool allocator.

However, since 590258b3814, the default heap-add with CONFIG_POSIX_API
has been reduced from 1 kiB which causes tests/posix/semaphores to fail
due to NULL being returned from a call to k_calloc().

Create a minimal heap-add for the POSIX_SEMAPHORES Option Group.

This can be removed at a future date if semaphores are changed to use
a pooled allocator and fixed-size name, rather than heap allocation.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: semaphores: reduce execution time by 5s

There is no reason to use a 5s wait time in tests. This can add up quite
significantly across multiple platforms that execute in real-time under
qemu (or even just on real hardware).

Speedup observed with qemu_cortex_a53/qemu_cortex_a53/smp

Before:
```
START - test_named_semaphore
 PASS - test_named_semaphore in 5.688 seconds
```

After:
```
START - test_named_semaphore
 PASS - test_named_semaphore in 0.783 seconds
```

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: semaphores: coalesce string constants

Just a formatting change, in a separate commit.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
* rebased
* posix: semaphores: use a default minimal heap-add for semaphores
* tests: posix: semaphores: reduce execution time by 5s
* tests: posix: semaphores: coalesce string constants
* tests: posix: semaphores: ensure test is not skipped

cfriedt commented:
* ensure adding `WAIT_TIME_MS` to `abstime` results in a valid `timespec`

ycsin commented on tests/posix/semaphores/src/main.c:
```diff
@@ -51,11 +54,10 @@ static void semaphore_test(sem_t *sem)
 	zassert_equal(clock_gettime(CLOCK_REALTIME, &abstime), 0, "clock_gettime failed");
 
-	abstime.tv_sec += 5;
+	abstime.tv_sec += WAIT_TIME_MS / MSEC_PER_SEC;
```
Is this supposed to be in the second commit?

cfriedt replied:
Fixed!
